src/services/TestTypesService.ts

```diff
@@ -1,220 +1,321 @@
-                }
-
-                if (testCodes.length > 1) {
-                    console.error("More than one testType was retrieved");
-                    throwInternalServerError();
-                }
-
-                const response: any = {
-                    id: testType.id
-                };
-
-                filterExpression.fields // Iterate through filterExpression's fields and populate them in the response
-                    .forEach((field: keyof TestCode) => {
-                        response[field] = testCodes[0][field];
-                    });
-
-                // Populating testTypeClassification that is found in testType, not testCode
-                if (filterExpression.fields.includes("testTypeClassification")) {
-                    response.testTypeClassification = testType.testTypeClassification;
-                }
-                return response;
-            });
-    }
+        return this.sort(filteredTestType);
+      })
+      .catch((error) => {
+        if (!(error instanceof HTTPError)) {
+          console.error(error);
+          error.statusCode = 500;
+          error.body = "Internal Server Error";
+        }
 
-    /**
-     * returns null if the test-type was not found
-     * @param id
-     * @param testTypes
-     */
-    public findTestType({ id, testTypes }: { id: string; testTypes: any; }) {
-        for (const testType of testTypes) {
-            // for (let i = 0; i < testTypes.length; i++) {
-            //   const testType: any = testTypes[i];
-            if (testType.hasOwnProperty("nextTestTypesOrCategories")) {
-                const childrenTestType: any = this.findTestType({ id, testTypes: testType.nextTestTypesOrCategories });
-
-                if (childrenTestType != null) {
-                    return childrenTestType;
-                }
-            } else if (testType.hasOwnProperty("id") && testType.id === id) {
-                return testType;
-            }
+        throw new HTTPError(error.statusCode, error.body);
+      });
+  }
+
+  public getTestTypesById(id: string, filterExpression: any) {
+    return this.testTypesDAO
+      .getAll()
+      .then((data) => {
+        if (data.Count === 0) {
+          throwResourceNotFound();
         }
-        return null;
-    }
 
-    public sort(testTypes: ITestType[] | any) {
-        // Pass by value
-        const testTypeArray = testTypes;
+        return data.Items;
+      })
+      .then((testTypes) => {
+        return this.findTestType({ id, testTypes });
+      })
+      .then((testType) => {
+        if (testType === null) {
+          throwResourceNotFound();
+        }
+        let testCodes: TestCode[] = testType.testCodes;
+
+        testCodes = testCodes
+          .filter((testCode) => {
+            // filter by vehicleType if present in DB, otherwise skip
+            return testCode.forVehicleType
+              ? this.fieldInFilterExpressionMatchesTheOneInTestCode(
```

Comment:
I think we could get rid off the ternary operator here since we always seem to be returning true and possibly use the shortcircuit operator when we chain the filters

return testCode. forVehicleType && this. fieldInFilterExpressionMatchesTheOneInTestCode(...)